### PR TITLE
Add EIP-5380: EIP-721 Entitlement Extension

### PR DESCRIPTION
**Pull request by Pandapip1**
Like EIP-4907 but completely different.

### CODE REVIEW

**Comment by Pandapip1** [comment]
> file EIPS/eip-xxxx.md is not a valid filename, but this error has been ignored due to editor approvals

*gulp* I guess it's lucky there was an eipw error then...

File: EIPS/eip-5380.md
Comment (MicahZoltu) [inline EIPS/eip-5380.md:7]:
```suggestion
status: Draft
```
EIPs all must go through the process, no special treatment for any reason (too fuzzy, too much contention).  Note, you can proceed through all the steps except last call as fast as you can get editor approval/merge.

File: EIPS/eip-5380.md
Comment (MicahZoltu) [inline EIPS/eip-5380.md:6]: EIPs should generally have their own discussion link to avoid confusion.  I recommend linking to this in the OP of that discussion topic, but we should keep the topics separate so there isn't confusion about what EIP is being discussed.

File: EIPS/eip-5380.md
Comment (MicahZoltu) [inline EIPS/eip-5380.md:5]: @Pandapip1 Since you are the one authoring this, I recommend adding yourself as an author, otherwise you'll need to wait to merge this as draft (and any subsequent steps in the process) until an author approves.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:5]: I thought I had done that.
```suggestion
author: Anders (@0xanders), Lance (@LanceSnow), Shrug <shrug@emojidao.org>, Tim Daubenschütz (@TimDaub), Pandapip1 (@Pandapip1)
```

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:6]:
```
discussions-to: https://ethereum-magicians.org/t/pr-5380-eip-4907-alternative-design/10190
```

**Comment by abcoathup** [comment]
Does the title need to reference EIP4907?  Could it just be: **ERC-721 User and Expires extension**

**Comment by TimDaub** [comment]
> @TimDaub you wanted this, so here you have it. Any other changes you would have made?

Hey @Pandapip1 thanks so much, I'll try to get this reviewed asap.

**Comment by Pandapip1** [comment]
> Hey @Pandapip1 thanks so much, I'll try to get this reviewed asap.

No need to do that. I've added myself as an author so I can get this through all the stages ASAP. We're just waiting for the other EIP editors here.

**Comment by Pandapip1** [comment]
Just realized that if I'm going to be a co-author of this, I might as well make the changes *I* would have made. Moving to draft as I make changes.

File: EIPS/eip-5380.md
Comment (Pandapip1) [inline EIPS/eip-5380.md:26]:
```suggestion
### `IERC5380`
```

File: EIPS/eip-5380.md
Comment (Pandapip1) [inline EIPS/eip-5380.md:43]:
```suggestion
    /// @notice             Get the maximum number of users that can receive this entitlement
```

File: EIPS/eip-5380.md
Comment (Pandapip1) [inline EIPS/eip-5380.md:58]:
```suggestion
### `IERC5380Enumerable`
```

**Comment by Pandapip1** [comment]
@TimDaub would you mind looking at this? It's a lot different than EIP-4907 now.

**Comment by TimDaub** [comment]
Hey sorry for taking suuuuper long to get to this. @Pandapip1 thanks so much for taking this on, it looks actually really interesting what you're doing here. I think it is in line with what has been outlined in the Decentralized Society paper, specifically this section:

Screenshot 2022-08-17 at 16 10 55

What do you think about listing a bunch of entitlement types and adding an enum-type structure that captures the entitlement that can be sent to a user?

## References:

- 1: https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4105763

**Comment by Pandapip1** [comment]
> What do you think about listing a bunch of entitlement types and adding an enum-type structure that captures the entitlement that can be sent to a user?

Do you mean add a metadata extension? That does seem like a good idea.

File: EIPS/eip-5380.md
Comment (xinbenlv) [inline EIPS/eip-5380.md:20]: What's an on-chain TCG?.

File: EIPS/eip-5380.md
Comment (xinbenlv) [inline EIPS/eip-5380.md:34]: Update event name for better readabiliy

```suggestion
    event EntitlementChanged(address indexed user, address indexed contract, uint256 indexed tokenId);
```

File: EIPS/eip-5380.md
Comment (xinbenlv) [inline EIPS/eip-5380.md:41]: I suggest add a `bytes calldata data` for future extension or reason of entitlement

File: EIPS/eip-5380.md
Comment (xinbenlv) [inline EIPS/eip-5380.md:83]: [Optional] If this EIP adds `uint256 amount` in the interfaces, it could have the benefit of supporting *both* EIP-721 and EIP-1155, potentially helpful for its adoption.

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:20]: TCG = trading card game. I'll define it.

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:41]: I'll add an additional function that supports it. I'm not a huge fan but I can see possible use cases.

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:83]: The issue here is supporting `maxEntitlements`. I initially was trying to do it in a cross-token manner, but it just didn't particularly work well. It's not clear who loses what when the balance changes without going to zero.

**Comment by TimDaub** [comment]
> Do you mean add a metadata extension? That does seem like a good idea.

Yeah but preferably it's onchain and e.g. it allows someone to say: "I, address B, entitle address A to 'usus' Property p", where 'usus' is the entitlement and there can be many different ones all encoded in an enum.

**Comment by Pandapip1** [comment]
Hmm... what about being able to have entitlements extend from other entitlements (i.e. the default address is the person with the `usus` entitlement)?

File: EIPS/eip-5380.md
Comment (SamWilsn) [inline EIPS/eip-5380.md:64]: Generally I recommend including the interface id explicitly in the EIP. Helps to catch bugs when copying the interface.

File: EIPS/eip-5380.md
Comment (SamWilsn) [inline EIPS/eip-5380.md:83]: Same here: include the interface id.

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:33]: This reads wrong to me. Would you mind explaining why you prefer adding a `has`?

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:64]: Personally, I prefer keeping it this way while the EIP is in draft. When it moves to review I'd agree that it's probably best to explicitly state it.

File: EIPS/eip-5380.md
Comment (Pandapip1) [reply EIPS/eip-5380.md:33]: `changes` here is used in the 3rd person present.

Comment (SamWilsn) [reply EIPS/eip-5380.md:33]: To confirm my understanding of your intent, below is what I _think_ you're trying to say (worded slightly differently):

> `EntitlementChanged` is emitted when the amount of entitlement assigned to a user changes.

If my understanding is correct, I think "a user" is an incomplete adjective clause describing "amount of entitlement". What I would suggest is "the amount of entitlement that a user has", or omitting the optional relative pronoun `that`, "the amount of entitlement a user has".

I am not suggesting you use a compound verb.

Comment (Pandapip1) [reply EIPS/eip-5380.md:33]: That makes sense.

**Comment by TimDaub** [comment]
> Hmm... what about being able to have entitlements extend from other entitlements (i.e. the default address for an entitlement extending from the `usus` entitlement is the person with the `usus` entitlement)?

I think this standards usefulnes can come from doing some research on the possible existing entitlements that exist, naming them appropriately and then recommending them as an enum:

- usus
- abusus
- etc.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**
